include/seastar/json/json_elements.hh

```diff
@@ -367,6 +370,47 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
     };
 }
 
+/*!
+ * \brief consume values from a coroutine generator \c gen and encode the items using the \c fun mapping function
+ * onto the \c out output_stream as a json array.
+ */
+template<typename T, typename Func>
+requires requires (T val, Func aa, output_stream<char> s) { { formatter::write(s, aa(val)) } -> std::same_as<future<>>; }
+future<> generate_array_element(output_stream<char>& out, coroutine::experimental::generator<T>& gen, Func fun) {
```

Comment:
Looking at std::generator, it implements std::ranges::view_interface, so you normally wouldn't consume a generator, you consume a view. The user can pass anything that's a view (like std::views::iota), or a generator if they like.

We should have the concept of a streaming view: something that has std::optional\<T> operator()().

Reply:
> Why generate_array_element? It's writing the whole array, no?

I think `element` here means `json element`, which is still confusing, but I don't have a better idea (maybe generate_json_array_type?), i.e. `{ "some_array": generage_array_element() }`

Reply:
> Why generate_array_element? It's writing the whole array, no?

The name isn't great, it's not generating an element in an array, but rather the whold array as a json element.

Reply:
> Why an extra mapping? Seems like it can be applied outside.

Hmm, the interface evolved from stream_range_as_array hence the mapping function, but I agree that using a generator indeed allows the user to compose the mapper externally as part of the generator.

Reply:
> Looking at std::generator, it implements std::ranges::view_interface, so you normally wouldn't consume a generator, you consume a view. The user can pass anything that's a view (like std::views::iota), or a generator if they like.
> 
> We should have the concept of a streaming view: something that has std::optional<T> operator()().

But it's an async generator so the view should have `future<std::optional<T>> operator()()`

Reply:
Yes, I forgot the future there.

Reply:
The mapping is still there.

Reply:
It's simple on paper but the optional result needs to be moved too so it needs to hold a std::unique_ptr<json_base>

Reply:
How can fun return json_base if the generator can't?

Reply:
> How can fun return json_base if the generator can't?

It doesn't.  It returns an object derived from json_base that isn't moved further, but rather "serialized" onto the json stream.

I think I have a solution by allowing the implementing a noexcept move constructor for the test's object_json
and defining the generator as returning object_json with the right contraints.

I'll post a new version

Reply:
Done in [1ec981a](https://github.com/scylladb/seastar/commit/1ec981a2ac274ce5478eed374c0ea40767d1b7c4)